### DISCUSSION TITLE
feat: Type Definition for the data property of QueryResult

### DISCUSSION THREAD

**Discussion by sshanzel** [TanStack/query#1267] [Ideas]
First, thank you for this wonderful library. Second, I'd love to contribute just let me know where I can start 🎉 

**Describe the ~~bug~~ feature request**
Currently, the `type` for the `getQueryData` can only be defined by a generic parameter. I think we can also use a generic `type` that extends the `typeof` of a custom hook where we originally used the `key` which will infer the `TResult` of the `QueryResult`. So that we can also pass the `typeof` of a custom hook and the type will automatically be what is desired.

I am fairly new with the library so please kindly inform me if I misunderstood the correct implementation.

**To Reproduce**
Steps to reproduce the behavior - this is how I use `react-query`:
![image](https://user-images.githubusercontent.com/13744167/98465165-c9628500-21e0-11eb-8048-35dd1aabfc20.png)

Then whenever I want to fetch the cached data which the `SAMPLE` key was used. I do this:
![image](https://user-images.githubusercontent.com/13744167/98465211-10e91100-21e1-11eb-8a9f-9a79dcc9aed2.png)

In my humble opinion, I do feel that there is a greater margin for mistyping the actual definition of the desired custom hook when type inference is being utilized (like the example above with `useSample` hook. I am actually a victim of this problem as I didn't realize the `getQueryData` returns the `data` property already so I defined my `getQueryData` execution with the `ReturnType<typeof useSample>` without even checking the documentation of the actual return `type`).

**Expected behavior**
I defined a type that infers the generic parameter which resulted to this:
![image](https://user-images.githubusercontent.com/13744167/98465433-7558a000-21e2-11eb-8b96-65756ead9e1b.png)
Instead of returning the type `QueryResult` it inferred the type of the `data` property which I think is appropriate with the current case for `getQueryData`.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [Windows]
 - Browser: [Chrome]
 - Version: [86.0.4240.183]

**Smartphone (please complete the following information):**
 - Browser: [e.g. stock browser, safari]
 - Version: [e.g. 22]

**Additional context**
I can immediately request a PR if this is deemed appropriate 💯 

Related [Issue](https://github.com/tannerlinsley/react-query/issues/389)

**Comment by TkDodo** [comment]
Not sure what you would want to add to the library. I think the types are technically correct because react-query cannot know what is in the cache, thus the type is `unknown`. I can put an arbitrary value into the cache at any time. We are also not forced to use custom hooks, so I don't quite get it. Can you elaborate a bit on what you would want to do to improve the types please?

**Comment by sshanzel** [comment]
I guess I was just wanting it to be how Redux infers the `actions` once the `type` property is defined which is if we're talking about `react-query` it would be when you define the `key` that it will be able to infer the type on where it was used (like on custom hooks). 

No worries, I have just created a wrapper on my project for the `getQuery` and `getQueryData` to achieve the screenshots I have attached above.

**Comment by tannerlinsley** [comment]
Seeing how this is not a bug per se, going to convert to a discussion.

**Comment by eddietex** [comment]
I just had the same thought while using `getQueryData`: "Why is this not inferring the type based on the key I'm providing?"

Not sure how easy to accomplish this is though. I'd require a setup up-front to tell the library what are the types for each key.

Maybe the best solution for this is to just create a helper function that knows of that type, like @sshanzel mentioned.

Curious if others also had the same thought, and what their solution for this was.

**Comment by roblabat** [comment]
Hi @eddietex, 

I have the same thought about getQueryData. I implemented a solution on my project based on types override of the lib typings and I think it could be implemented in the lib easily. I opened a discussion on my suggestion, feel free to comment. [https://github.com/TanStack/query/discussions/6105](https://github.com/TanStack/query/discussions/6105)
Does this would solve your type inference issue ?
